backfiller/backfiller/main.py

```diff
@@ -129,159 +114,308 @@ def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
-	prom.start_http_server(metrics_port)
-
-	if backdoor_port:
-		gevent.backdoor.BackdoorServer(('127.0.0.1', backdoor_port), locals=locals()).start()
-
-	logging.info('Starting backfilling {} with {} as variants to {}'.format(stream, ', '.join(variants), base_dir))
-
-	fill_start = datetime.datetime.now()
-	full_fill_start = fill_start
+	def stop(self):
+		"""Tell the worker to shut down"""
+		self.logger.info('Stopping')
+		self.stopping.set()
 
+	def backfill(self, hours, segment_order='random', recent_cutoff=60):
+		"""Backfill from remote node.
+	
+		Backfill from node/stream/variants to base_dir/stream/variants for each
+		hour in hours.
+	
+		Keyword arguments:
+		recent_cutoff -- Skip backfilling segments younger than this number of
+			seconds to prioritise letting the downloader grab these segments."""
+	
+		for variant in self.variants:
+	
+			for hour in hours:
+	
+				self.logger.info('Backfilling {}/{}'.format(variant, hour))
+	
+				local_segments = set(list_local_segments(self.base_dir, self.stream, variant, hour))
+				remote_segments = set(list_remote_segments(self.node, self.stream, variant, hour))
+				missing_segments = list(remote_segments - local_segments)
+	
+				# randomise the order of the segments to reduce the chance that
+				# multiple workers request the same segment at the same time
+				random.shuffle(missing_segments)
+	
+				for missing_segment in missing_segments:
 	
-	backfill(base_dir, stream, variants, 3, nodes=nodes)
+					if self.stopping.is_set():
+						return
 
-	backfill(base_dir, stream, variants, nodes=nodes)
+					path = os.path.join(self.stream, variant, hour, missing_segment)
 	
-	# I'm sure there is a module that does this in a more robust way 
-	# but I understand this and it gives the behaviour I want
-	while True:
+					# test to see if file is a segment and get the segments start time
+					try:
+						segment = common.parse_segment_path(path)
+					except ValueError as e:
+						self.logger.warning('File {} invaid: {}'.format(path, e))
+						continue
+	
+					#to avoid getting in the downloader's way ignore segments less than recent_cutoff old
+					if datetime.datetime.utcnow() - segment.start < datetime.timedelta(seconds=recent_cutoff):
+						self.logger.debug('Skipping {} as too recent'.format(path))
+						continue
+	
+					get_remote_segment(self.base_dir, self.node, self.stream, variant, hour, missing_segment)
+				self.logger.info('{} segments in {}/{} backfilled'.format(len(missing_segments), variant, hour))
+
+	def run(self):
+		self.logger.info('Starting')
+
+		while not self.stopping.is_set():
+
+			try:
+				self.backfill(list_hours(self.node, self.stream, self.variants, self.start))
+				self.failures = 0 #reset failure count on a successful backfill
+				if not self.run_once:
+					self.stopping.wait(common.jitter(self.WAIT_INTERVAL))
+
+			except Exception:
+				self.failures += 1
+				delay = common.jitter(self.RETRY_INTERVAL * 2**self.failures)
```

Comment:
this expontential backoff needs an upper limit, or it will take forever to retry after a few rounds. I suggest something like one minute (reached after 4 failures).

Reply:
Added a limit of 4 tries